pkg/plugins/golang-config-gen/v1/api.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/pflag"
+
+	"sigs.k8s.io/kubebuilder/v3/pkg/config"
+	"sigs.k8s.io/kubebuilder/v3/pkg/machinery"
+	"sigs.k8s.io/kubebuilder/v3/pkg/model"
+	"sigs.k8s.io/kubebuilder/v3/pkg/model/resource"
+	"sigs.k8s.io/kubebuilder/v3/pkg/plugin"
+	"sigs.k8s.io/kubebuilder/v3/pkg/plugin/util"
+	goPlugin "sigs.k8s.io/kubebuilder/v3/pkg/plugins/golang"
+	"sigs.k8s.io/kubebuilder/v3/pkg/plugins/golang-config-gen/v1/scaffolds"
+	"sigs.k8s.io/kubebuilder/v3/pkg/plugins/internal/cmdutil"
+	"sigs.k8s.io/kubebuilder/v3/plugins/addon"
+)
+
+const (
+	// KbDeclarativePatternVersion is the sigs.k8s.io/kubebuilder-declarative-pattern version
+	// (used only to gen api with --pattern=addon)
+	// TODO: remove this when a better solution for using addons is implemented.
+	KbDeclarativePatternVersion = "b84d99da021778217217885dd9582ed3cc879ebe"
+
+	// defaultCRDVersion is the default CRD API version to scaffold.
+	defaultCRDVersion = "v1"
+)
+
+// DefaultMainPath is default file path of main.go
+const DefaultMainPath = "main.go"
```

Comment:
Some comments, otherwise relly nice we see it moving forward.

We cannot drop do the support to the plugins, so IMO we need to allow we use it opt-in opt which we will be to achieve with https://github.com/kubernetes-sigs/kubebuilder/issues/2015. So, can we:
-  Have a plugin that only scaffold what is of its domain of responsibility (I d not think that. we should triplicated the files here )
- Then, we can create a testdata for go v3 where we will scaffold the project, remove the config dir and then add the config-gen for we start to try to make it backwards compatible with the plugin until we are able to push/merge https://github.com/kubernetes-sigs/kubebuilder/issues/2015?

Reply:
* Before this goes into master, I agree. I'd like to see this refactored into something like `config-gen/v1-alpha`, where `config/v3` is the default `config/` directory plugin.
* Sure